### DISCUSSION TITLE
preHandler vs onRequest

### DISCUSSION THREAD

**Discussion by gajus** [fastify/fastify#3772] [Q&A] [answered]
It is not entirely clear when one would use one or another.

**Comment by Eomm** [comment, marked as answer]
The main difference is that the `onRequest` has not the body payload parsed, the `preHandler` has it 
https://www.fastify.io/docs/latest/Reference/Lifecycle/#lifecycle


So, do you need to manipulate the body? Use `preHandler`
Do you need to know that a request has come into the server? Use `onRequest`

**Reply by jsumners** [reply]
Some authentication actions should be performed _before_ any request parsing.

**Reply by steven-gpa** [reply]
Could you please extend this comment?

**Reply by Eomm** [reply]
Example:

The client sends a 10MB payload.

If the client is not authorized and you check the authentication on `preHandler`, your server will read and store in memory 10MB per request even if the client is not authenticated.

If the client does the same thing, but you are using `onRequest`, the server did not read the request's body payload, so it is not allocating 10MB.

So technically, an unauthorized user could DOS your server by exhausting your server's resources

**Reply by steven-gpa** [reply]
Thank you so much for your prompt response and the example provided 😃 👍

**Reply by steven-gpa** [reply]
I was analyzing the https://github.com/fastify/fastify-auth package, but the code example (see below) of the implementation was passing the `fastify.auth`  in the `preHandler` property, so based on the above scenario, it will be better to implement the `fastify.auth` in the `onRequest` property.

```js
fastify.route({
  method: 'GET',
  url: '/run-all',
  preHandler: fastify.auth([ // ---> onRequest rather than preHandler
    (request, reply, done) => { console.log('executed 1'); done() },
    (request, reply, done) => { console.log('executed 2'); done() },
    (request, reply, done) => { console.log('executed 3'); done(new Error('you are not authenticated')) },
    (request, reply, done) => { console.log('executed 4'); done() },
    (request, reply, done) => { console.log('executed 5'); done(new Error('you shall not pass')) }
  ], { run: 'all' }),
  handler: (req, reply) => { reply.send({ hello: 'world' }) }
})
```

**Reply by Uzlopak** [reply]
The problem is, that not all authentication methods are only based on http headers. E.g. OAuth2 needs the body. So for that the body has to be parsed.

**Reply by steven-gpa** [reply]
Gotcha, thanks again for the prompt response 🙇

**Reply by Uzlopak** [reply]
Created https://github.com/fastify/fastify-auth/pull/180 ;)

**Reply by steven-gpa** [reply]
Awesome!! 🕺 , thanks @Uzlopak !! 🥇
